openflow15/match.go

```diff
@@ -0,0 +1,2200 @@
+	ipSrcField.Ipv6Src = ipSrc
+	f.Value = ipSrcField
+	f.Length = uint8(ipSrcField.Len())
+
+	// Add the mask
+	if ipSrcMask != nil {
+		mask := new(Ipv6SrcField)
+		mask.Ipv6Src = *ipSrcMask
+		f.Mask = mask
+		f.HasMask = true
+		f.Length += uint8(mask.Len())
+	}
+
+	return f
+}
+
+// IPV6_FLABEL
+type Ipv6FLabelField struct {
+	FLabel uint32
+}
+
+func (m *Ipv6FLabelField) Len() uint16 {
+	return 4
+}
+
+func (m *Ipv6FLabelField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, m.Len())
+
+	binary.BigEndian.PutUint32(data[0:], m.FLabel)
+	return
+}
+
+func (m *Ipv6FLabelField) UnmarshalBinary(data []byte) error {
+	if len(data) < int(m.Len()) {
+		return errors.New("The byte array has wrong size to unmarshal Ipv6FLabelField message")
+	}
+	m.FLabel = binary.BigEndian.Uint32(data[0:])
+	return nil
+}
+
+func NewIpv6FLabelField(flabel uint32, flabelMask *uint32) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IPV6_FLABEL
+	f.HasMask = true
+
+	flabelField := new(Ipv6FLabelField)
+	flabelField.FLabel = flabel
+	f.Value = flabelField
+	f.Length = uint8(flabelField.Len())
+
+	// Add the mask
+	if flabelMask != nil {
+		mask := new(Ipv6FLabelField)
+		mask.FLabel = *flabelMask
+		f.Mask = mask
+		f.HasMask = true
+		f.Length += uint8(mask.Len())
+	}
+	return f
+}
+
+// IPV6_DST field
+type Ipv6DstField struct {
+	Ipv6Dst net.IP
+}
+
+func (m *Ipv6DstField) Len() uint16 {
+	return 16
+}
+func (m *Ipv6DstField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 16)
+	copy(data, m.Ipv6Dst)
+	return
+}
+
+func (m *Ipv6DstField) UnmarshalBinary(data []byte) error {
+	m.Ipv6Dst = make([]byte, 16)
+	copy(m.Ipv6Dst, data)
+	return nil
+}
+
+// Return a MatchField for ipv6 dest addr
+func NewIpv6DstField(ipDst net.IP, ipDstMask *net.IP) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IPV6_DST
+	f.HasMask = false
+
+	ipDstField := new(Ipv6DstField)
+	ipDstField.Ipv6Dst = ipDst
+	f.Value = ipDstField
+	f.Length = uint8(ipDstField.Len())
+
+	// Add the mask
+	if ipDstMask != nil {
+		mask := new(Ipv6DstField)
+		mask.Ipv6Dst = *ipDstMask
+		f.Mask = mask
+		f.HasMask = true
+		f.Length += uint8(mask.Len())
+	}
+
+	return f
+}
+
+// IP_ECN field
+type IpEcnField struct {
+	IpEcn uint8
+}
+
+func (m *IpEcnField) Len() uint16 {
+	return 1
+}
+func (m *IpEcnField) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 1)
+
+	data[0] = m.IpEcn
+	return
+}
+func (m *IpEcnField) UnmarshalBinary(data []byte) (err error) {
+	m.IpEcn = data[0]
+	return
+}
+
+// Return a MatchField for vlan id matching
+func NewIpEcnField(vlanPcp uint8) *MatchField {
+	f := new(MatchField)
+	f.Class = OXM_CLASS_OPENFLOW_BASIC
+	f.Field = OXM_FIELD_IP_ECN
+	f.HasMask = false
+
+	vlanPcpField := new(IpEcnField)
+	f.Value = vlanPcpField
+	f.Length = uint8(vlanPcpField.Len())
+
+	return f
+}
+
+// IP_PROTO field
+type IpProtoField struct {
+	protocol uint8
```

Comment:
Not introduced by you, but it should be better to use a public property for such a library, then it is convenient for callers to reset the values if necessary.

Reply:
Done.